### PR TITLE
Compilation_unit optimisations

### PR DESCRIPTION
**Pull request by mshinwell**
There is a problem with build artifact size at tag `4.14.0-0`.  All `*.cm*` files are about 25% bigger on the JS tree build under Closure.  Much of this comes from the types used to represent `cmi` and `cmx` imports; we had to change these for the functor packs work (and in any case they weren't very efficient, marshalling as lists rather than arrays for example).  We think there might also be some other changes causing `cmi` files to be larger, but will look at those later.

We have two changes to make to rectify this problem.  This is the first, which uses an `Obj` trick to significantly improve the memory layout of values of type `Compilation_unit.t` (and makes sure that sharing isn't broken by functions like `full_path_as_string`).  It seems likely that we will be able to remove this in due course given work on similar kinds of unboxing going on elsewhere.

The second replaces the tuple types that are scattered around the codebase with a single `Import_info.t` type which is abstract and whose memory layout is much better.  This will be posted as a separate PR.

Joint work with @xclerc .

Includes #1034.

### CODE REVIEW

File: ocaml/utils/compilation_unit.ml [hunk 1]
```diff
@@ -123,146 +139,160 @@ end = struct
...
+  assert (tag = 0 || tag = Obj.string_tag);
+  if tag <> 0
+  then Prefix.empty
+  else
+    let with_prefix = Sys.opaque_identity (Obj.obj t : with_prefix) in
+    with_prefix.for_pack_prefix
 
 let create for_pack_prefix name =
-  if not (Prefix.is_empty for_pack_prefix) then begin
+  if not (Prefix.is_empty for_pack_prefix)
```
Comment (xclerc) [inline ocaml/utils/compilation_unit.ml:217]: (It is certainly cheap, but we are testing
whether the prefix is empty twice.)

File: ocaml/utils/compilation_unit.ml [hunk 2]
```diff
@@ -281,62 +312,61 @@ let which_cmx_file desired_comp_unit ~accessed_by : Name.t =
...
     match_components ~current:(full_path accessed_by)
       ~desired:(full_path desired_comp_unit)
 
-let print_name ppf t =
-  Format.fprintf ppf "%a" Name.print t.name
+let print_name ppf t = Format.fprintf ppf "%a" Name.print (name t)
 
 let full_path_as_string t =
-  Format.asprintf "%a" print t
+  if Prefix.is_empty (for_pack_prefix t)
```
Comment (xclerc) [inline ocaml/utils/compilation_unit.ml:350]: (Maybe leave a CR stating that
non-empty-prefix path breaks
sharing.)

File: ocaml/utils/compilation_unit.ml [hunk 3]
```diff
@@ -123,146 +139,160 @@ end = struct
...
 let dummy = create Prefix.empty (Name.of_string "*none*")
 
 let predef_exn = create Prefix.empty Name.predef_exn
 
-let name t = t.name
-
 let name_as_string t = name t |> Name.to_string
 
-let for_pack_prefix t = t.for_pack_prefix
+let with_for_pack_prefix t for_pack_prefix = create for_pack_prefix (name t)
```
Comment (xclerc) [inline ocaml/utils/compilation_unit.ml:250]: (Now checks whether the passed prefix
is valid, probably a good thing. Nothing
in the code seems to try and build an
invalid prefix - I am thinking of something
like `*predef*`.)

**Review by xclerc — APPROVED** [review]
Using `Obj` here will be controversial,
but the size improvements are quite
probably worth the risk.

All comments below are minor and
can be safely ignored.

File: ocaml/utils/compilation_unit.ml (hunk 3 above)
Comment (mshinwell) [reply ocaml/utils/compilation_unit.ml:250]: This indeed should be ok.

**Comment by stedolan** [comment]
What is going on here? It sounds from the description that a large proportion of `*.cm*` is instances of `Compilation_unit.`, otherwise shrinking them wouldn't have much effect. But that can't be right?

**Comment by mshinwell** [comment]
From what I recall, there are a lot of values of that type in the `.cm*` files, in particular in the import lists.

**Pull request merged**